asm/opcode.go

```diff
@@ -71,24 +71,29 @@ func (cls Class) isJumpOrALU() bool {
 //
 // The encoding varies based on a 3-bit Class:
 //
-//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
-//	           ???           | CLS
+//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+//	                          ???                            | CLS
 //
 // For ALUClass and ALUCLass32:
 //
-//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
-//	           OPC         |S| CLS
+//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+//	             0                 |           OPC         |S| CLS
 //
 // For LdClass, LdXclass, StClass and StXClass:
 //
-//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
-//	        0      | MDE |SIZ| CLS
+//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+//	                       0                       | MDE |SIZ| CLS
+//
+// For StXClass where MDE == AtomicMode:
+//
+//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+//	              0              |    ATOMIC OP    | MDE |SIZ| CLS
 //
 // For JumpClass, Jump32Class:
 //
-//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
-//	        0      |  OPC  |S| CLS
-type OpCode uint16
+//	7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+//	                       0                       |  OPC  |S| CLS
+type OpCode uint32
```

Comment:
Nit: we could stick to 16bit now? Though 32bit is also nice, because we will never run out of space with that I think.

Reply:
Going to 32bit doesn't change size of Instruction, so this is probably fine.

Reply:
We still need the 9th bit. For Atomic ops, we would use the lower 2 bytes, and the lowest bit of the upper 2 bytes.
So the mask is `0x0001FFFF`. 

We could likely avoid it if we take advantage of the fact that for atomics only 4 ALU instructions are allowed and renumber everything so it fits in in a single byte. But having the larger opcode is the least intrusive.